Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp

```diff
@@ -942,32 +942,38 @@ bool webKitSrcPassedCORSAccessCheck(WebKitWebSrc* src)
     return members->didPassAccessControlCheck;
 }
 
-CachedResourceStreamingClient::CachedResourceStreamingClient(const WebKitWebSrc& src, ResourceRequest&& request, unsigned requestNumber)
+CachedResourceStreamingClient::CachedResourceStreamingClient(WebKitWebSrc* src, ResourceRequest&& request, unsigned requestNumber)
     : m_requestNumber(requestNumber)
-    , m_src(GST_ELEMENT_CAST(&src))
     , m_request(WTFMove(request))
 {
+    g_weak_ref_init(&m_src, src);
 }
 
-CachedResourceStreamingClient::~CachedResourceStreamingClient() = default;
+CachedResourceStreamingClient::~CachedResourceStreamingClient()
+{
+    g_weak_ref_clear(&m_src);
+}
 
 void CachedResourceStreamingClient::checkUpdateBlocksize(unsigned bytesRead)
 {
     ASSERT(isMainThread());
-    WebKitWebSrc* src = WEBKIT_WEB_SRC(m_src);
-    GstBaseSrc* baseSrc = GST_BASE_SRC_CAST(src);
+    GRefPtr<WebKitWebSrc> src(WEBKIT_WEB_SRC_CAST(g_weak_ref_get(&m_src)));
```

Comment:
g_weak_ref_get returns a reference, so you need to adopt it. We already have GWeakPtr in WTF, but I guess we need thread safety here, so maybe we can add GThreadSafeWeakPtr to wrap GWeakRef and make it more difficult (or impossible) to make this kind of mistakes.

Reply:
TIL `WTF::GWeakPtr`! ThreadSafe as in check the ptr would be used from the same thread it was created from?

Reply:
I agree we should add `GThreadSafeWeakPtr`.

> ThreadSafe as in check the ptr would be used from the same thread it was created from?

No. (We should probably add an ASSERT() to the existing GWeakPtr to guarantee that, though. WTF::WeakPtr already has a similar check.)

GThreadSafeWeakPtr would be a weak pointer that guarantees you either atomically acquire a strong ref (preventing the object from being destroyed on another thread) or fail to do so (if the object is already being destroyed). It would need to be implemented using GWeakRef (as suggested by Carlos) directly rather than g_object_add_weak_pointer(). `GThreadSafeWeakPtr::get` would need to return a `GRefPtr` rather than a raw pointer.

Reply:
Also we would want to add a test to Tools/TestWebKitAPI/Tests/WTF/glib/GWeakPtr.cpp.

I don't think GThreadSafeWeakPtr would be too hard to create, but I also wouldn't want to make creating it a condition for fixing this bug.

Reply:
Oh, I see you've already done exactly this, and I just wasted my time by explaining what to do. Nice.

Reply:
> Also we would want to add a test to Tools/TestWebKitAPI/Tests/WTF/glib/GWeakPtr.cpp.

A quick test here would be nice to have, though.

Reply:
OK, I'll write a test tomorrow!